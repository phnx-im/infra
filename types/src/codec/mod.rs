// SPDX-FileCopyrightText: 2023 Phoenix R&D GmbH <hello@phnx.im>
//
// SPDX-License-Identifier: AGPL-3.0-or-later

use cbor::Cbor;
use error::CodecError;
use mls_assist::memory_provider::Codec;
use serde::{Serialize, de::DeserializeOwned};

mod cbor;
mod error;
mod persistence;
#[cfg(test)]
mod tests;

pub use error::Error;
pub use persistence::{BlobDecoded, BlobEncoded};

#[derive(Debug, Clone, Copy, PartialEq, Eq, Default)]
#[repr(u8)]
#[non_exhaustive]
pub enum PhnxCodec {
    #[cfg(test)]
    OlderTestVersion = 0,
    #[default]
    V1 = 1,
}

impl TryFrom<u8> for PhnxCodec {
    type Error = Error;

    fn try_from(value: u8) -> Result<Self, Self::Error> {
        match value {
            #[cfg(test)]
            0 => Ok(PhnxCodec::OlderTestVersion),
            1 => Ok(PhnxCodec::V1),
            _ => Err(Error::UnknownCodecVersion),
        }
    }
}

impl PhnxCodec {
    fn serialize_to_writer<T: Serialize>(
        &self,
        value: &T,
        writer: &mut impl std::io::Write,
    ) -> Result<(), Box<dyn std::error::Error + Send + Sync + 'static>> {
        // The first byte is always the codec version
        writer.write_all(&[*self as u8])?;
        match self {
            #[cfg(test)]
            PhnxCodec::OlderTestVersion => tests::Json::to_writer(value, writer)?,
            PhnxCodec::V1 => Cbor::to_writer(value, writer)?,
        }
        Ok(())
    }

    fn serialize<T: Sized + Serialize>(
        &self,
        value: &T,
    ) -> Result<Vec<u8>, Box<dyn std::error::Error + Send + Sync + 'static>> {
        let mut buf = Vec::new();
        self.serialize_to_writer(value, &mut buf)?;
        Ok(buf)
    }

    fn deserialize<T: DeserializeOwned>(
        &self,
        bytes: &[u8],
    ) -> Result<T, Box<dyn std::error::Error + Send + Sync + 'static>> {
        let res = match self {
            #[cfg(test)]
            PhnxCodec::OlderTestVersion => tests::Json::from_slice(bytes)?,
            PhnxCodec::V1 => Cbor::from_slice(bytes)?,
        };
        Ok(res)
    }

    pub fn to_vec<T>(value: &T) -> Result<Vec<u8>, Error>
    where
        T: Sized + Serialize,
    {
        let codec_version = PhnxCodec::default();
        let res = codec_version.serialize(value).map_err(|error| CodecError {
            codec_version,
            error,
        })?;
        Ok(res)
    }

    pub fn from_slice<T>(bytes: &[u8]) -> Result<T, Error>
    where
        T: DeserializeOwned,
    {
        let codec_version_byte = bytes.first().ok_or(Error::EmptyyInputSlice)?;
        let codec_version = PhnxCodec::try_from(*codec_version_byte)?;
        codec_version.deserialize(&bytes[1..]).map_err(|error| {
            CodecError {
                codec_version,
                error,
            }
            .into()
        })
    }
}

impl Codec for PhnxCodec {
    type Error = Error;

    fn to_vec<T>(value: &T) -> Result<Vec<u8>, Self::Error>
    where
        T: Sized + Serialize,
    {
        Self::to_vec(value)
    }

    fn from_slice<T>(bytes: &[u8]) -> Result<T, Self::Error>
    where
        T: DeserializeOwned,
    {
        Self::from_slice(bytes)
    }
}
