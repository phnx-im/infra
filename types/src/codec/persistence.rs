// SPDX-FileCopyrightText: 2025 Phoenix R&D GmbH <hello@phnx.im>
//
// SPDX-License-Identifier: AGPL-3.0-or-later

use serde::{Serialize, de::DeserializeOwned};
use sqlx::{Database, Decode, Encode, Type, encode::IsNull, error::BoxDynError};

use super::PhnxCodec;

pub struct BlobEncoded<T: Serialize>(pub T);

impl<DB: Database, T: Serialize> Type<DB> for BlobEncoded<T>
where
    Vec<u8>: Type<DB>,
{
    fn type_info() -> <DB as Database>::TypeInfo {
        Vec::<u8>::type_info()
    }
}

impl<'q, DB: Database, T: Serialize> Encode<'q, DB> for BlobEncoded<T>
where
    Vec<u8>: Encode<'q, DB>,
{
    fn encode_by_ref(
        &self,
        buf: &mut <DB as Database>::ArgumentBuffer<'q>,
    ) -> Result<IsNull, BoxDynError> {
        let bytes = PhnxCodec::to_vec(&self.0)?;
        Encode::<DB>::encode(&bytes, buf)
    }
}

#[derive(Debug)]
pub struct BlobDecoded<T>(pub T);

impl<T> BlobDecoded<T> {
    pub fn into_inner(self) -> T {
        self.0
    }
}

impl<DB: Database, T: DeserializeOwned> Type<DB> for BlobDecoded<T>
where
    Vec<u8>: Type<DB>,
{
    fn type_info() -> <DB as Database>::TypeInfo {
        Vec::<u8>::type_info()
    }
}

impl<'q, DB: Database, T: DeserializeOwned> Decode<'q, DB> for BlobDecoded<T>
where
    for<'a> &'a [u8]: Decode<'a, DB>,
{
    fn decode(value: <DB as Database>::ValueRef<'q>) -> Result<Self, BoxDynError> {
        let bytes: &[u8] = Decode::<DB>::decode(value)?;
        PhnxCodec::from_slice(bytes)
            .map(BlobDecoded)
            .map_err(From::from)
    }
}
