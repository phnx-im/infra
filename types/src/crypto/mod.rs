// SPDX-FileCopyrightText: 2023 Phoenix R&D GmbH <hello@phnx.im>
//
// SPDX-License-Identifier: AGPL-3.0-or-later

//! A module that provides various traits and structs that allow the use of
//! cryptographic primitives such as AEAD, MACs and KDFs.
//!
//! TODO: Once const-generics allows the use of enums, we could get rid of a
//! number of structs and boilerplate code.
//! TODO: A proper RNG provider for use with all crypto functions that require
//! randomness, i.e. mainly secret and nonce sampling.
#![allow(unused_variables)]
use std::marker::PhantomData;

use hpke::{DecryptionKey, EncryptionPublicKey};
use serde::{Deserialize, Serialize};
use sha2::Sha256;
use thiserror::Error;
use tls_codec::{TlsDeserializeBytes, TlsSerialize, TlsSize};

use crate::{LibraryError, crypto::ear::EarEncryptable, messages::QueueMessage};

use self::{
    ear::{Ciphertext, EarDecryptable, keys::RatchetKey},
    errors::RandomnessError,
    hpke::{HpkeDecryptionKey, HpkeEncryptionKey},
    kdf::{KdfDerivable, keys::RatchetSecret},
};

pub use opaque::OpaqueCiphersuite;

/// This type determines the hash function used by the backend.
pub type Hash = Sha256;

pub mod ear;
pub mod errors;
pub mod hpke;
pub mod kdf;
pub mod mac;
pub mod opaque;
pub mod ratchet;
pub mod secrets;
pub(super) mod serde_arrays;
pub mod signatures;

pub type RatchetKeyUpdate = Vec<u8>;

#[derive(
    Debug, Clone, PartialEq, Eq, Serialize, Deserialize, TlsSerialize, TlsDeserializeBytes, TlsSize,
)]
pub struct RatchetEncryptionKey(EncryptionPublicKey);

impl RatchetEncryptionKey {
    #[cfg(any(test, feature = "test_utils"))]
    pub fn new_for_test(encryption_key: EncryptionPublicKey) -> Self {
        Self(encryption_key)
    }
}

#[derive(Clone, Serialize, Deserialize, sqlx::Type)]
#[sqlx(transparent)]
pub struct RatchetDecryptionKey(DecryptionKey);

impl RatchetDecryptionKey {
    pub fn generate() -> Result<Self, RandomnessError> {
        Ok(Self(DecryptionKey::generate()?))
    }

    pub fn encryption_key(&self) -> RatchetEncryptionKey {
        RatchetEncryptionKey(self.0.public_key().clone())
    }
}

#[derive(
    Debug, Clone, PartialEq, Eq, Serialize, Deserialize, TlsSerialize, TlsDeserializeBytes, TlsSize,
)]
pub struct ConnectionEncryptionKey {
    encryption_key: EncryptionPublicKey,
}

impl AsRef<EncryptionPublicKey> for ConnectionEncryptionKey {
    fn as_ref(&self) -> &EncryptionPublicKey {
        &self.encryption_key
    }
}

impl HpkeEncryptionKey for ConnectionEncryptionKey {}

#[derive(Clone, Serialize, Deserialize)]
pub struct ConnectionDecryptionKey {
    decryption_key: DecryptionKey,
}

impl AsRef<DecryptionKey> for ConnectionDecryptionKey {
    fn as_ref(&self) -> &DecryptionKey {
        &self.decryption_key
    }
}

impl HpkeDecryptionKey for ConnectionDecryptionKey {}

impl ConnectionDecryptionKey {
    pub fn generate() -> Result<Self, RandomnessError> {
        Ok(Self {
            decryption_key: DecryptionKey::generate()?,
        })
    }

    pub fn encryption_key(&self) -> ConnectionEncryptionKey {
        ConnectionEncryptionKey {
            encryption_key: self.decryption_key.public_key().clone(),
        }
    }
}

#[cfg(test)]
mod test {
    use crate::codec::PhnxCodec;

    use super::*;

    #[test]
    fn encryption_key_serde_codec() {
        let key = RatchetEncryptionKey::new_for_test(EncryptionPublicKey::from(vec![1, 2, 3]));
        insta::assert_binary_snapshot!(".cbor", PhnxCodec::to_vec(&key).unwrap());
    }

    #[test]
    fn encryption_key_serde_json() {
        let key = RatchetEncryptionKey::new_for_test(EncryptionPublicKey::from(vec![1, 2, 3]));
        insta::assert_json_snapshot!(key);
    }
}
