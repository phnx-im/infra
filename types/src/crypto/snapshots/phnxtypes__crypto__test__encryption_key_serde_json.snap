---
source: types/src/crypto/mod.rs
expression: "&key"
---
[
  1,
  2,
  3
]
